File: CRM-Server/app/models/opportunity.py
from sqlalchemy import Column, BigInteger, String, Integer, DateTime, Date, func, Index, ForeignKey, Numeric
from sqlalchemy.orm import relationship
from enum import Enum as PyEnum
from app.core.database import Base
from app.constants.approval_phase import ApprovalPhase


class PurchaseType(PyEnum):
    NEW = "NEW"
    RENEWAL = "RENEWAL"
    EXPANSION = "EXPANSION"


class LicenseType(PyEnum):
    SUBSCRIPTION = "SUBSCRIPTION"
    PERPETUAL = "PERPETUAL"


class OpportunityStatus(PyEnum):
    FOLLOWING = 0
    WON = 1
    LOST = 2


class OpportunityStage(Base):
    """
    销售阶段配置表

    支持团队隔离：不同团队可配置不同的销售阶段
    """
    __tablename__ = "crm_opportunity_stages"

    id = Column(BigInteger, primary_key=True, autoincrement=True, comment="主键")
    team_id = Column(BigInteger, nullable=True, index=True, comment="团队ID（NULL表示系统默认模板）")
    procurement_method_id = Column(BigInteger, nullable=True, comment="采购方式ID")
    stage_code = Column(String(100), nullable=False, comment="阶段编码")
    stage_name = Column(String(100), nullable=False, comment="阶段名称")
    win_probability = Column(Integer, nullable=False, default=0, comment="赢率（0-100）")
    sort_order = Column(Integer, nullable=False, default=0, comment="排序序号")
    description = Column(String(500), nullable=True, comment="阶段描述")
    is_active = Column(Integer, nullable=False, default=1, comment="是否启用：0:否, 1:是")
    is_default_start = Column(Integer, nullable=False, default=0, comment="是否默认起始阶段：0:否, 1:是")
    can_skip = Column(Integer, nullable=False, default=0, comment="是否可跳过：0:否, 1:是")

    created_time = Column(DateTime, nullable=False, default=func.now(), comment="创建时间")
    last_modified_time = Column(DateTime, nullable=False, default=func.now(), onupdate=func.now(), comment="最后修改时间")

    __table_args__ = (
        Index('idx_stage_code', 'stage_code'),
        Index('idx_sort_order', 'sort_order'),
        Index('idx_is_active', 'is_active'),
        Index('idx_procurement_method_id', 'procurement_method_id'),
        Index('idx_crm_opportunity_stages_team_id', 'team_id'),
        {'comment': '销售阶段配置表'}
    )


class Opportunity(Base):
    __tablename__ = "crm_opportunities"

    id = Column(BigInteger, primary_key=True, autoincrement=True, comment="主键")
    team_id = Column(BigInteger, nullable=False, index=True, comment="团队ID")
    opportunity_name = Column(String(255), nullable=False, comment="商机名称")
    customer_id = Column(BigInteger, ForeignKey('crm_customers.id', ondelete='CASCADE'), nullable=False, comment="关联客户ID")
    procurement_method_id = Column(BigInteger, nullable=True, comment="采购方式ID")
    current_stage_snapshot_id = Column(BigInteger, nullable=True, comment="当前阶段快照ID")
    current_stage_name = Column(String(100), nullable=True, comment="当前阶段名称")
    current_win_probability = Column(Integer, nullable=True, comment="当前阶段赢率")
    current_stage_entered_at = Column(DateTime, nullable=True, comment="当前阶段进入时间")
    total_amount = Column(Numeric(12, 2), nullable=False, comment="预计总金额")
    user_count = Column(Integer, nullable=False, comment="采购用户数")
    unit_price = Column(Numeric(10, 2), nullable=False, comment="标准单价（系统自动计算）")
    license_type = Column(String(20), nullable=False, comment="授权模式：SUBSCRIPTION:订阅, PERPETUAL:买断")
    subscription_years = Column(Integer, nullable=True, comment="订阅年限（订阅制时必填）")
    purchase_type = Column(String(20), nullable=False, comment="采购类型：NEW:新购, RENEWAL:续购, EXPANSION:增购")
    decision_maker_count = Column(Integer, nullable=True, comment="采购决策人数")
    expected_closing_date = Column(Date, nullable=False, comment="预计成交日期")
    procurement_stage_id = Column(BigInteger, nullable=True, comment="采购阶段ID（已废弃，保留用于兼容）")
    win_probability = Column(Integer, nullable=False, default=0, comment="当前阶段赢率（0-100）")
    owner_id = Column(String(100), nullable=False, comment="负责人系统用户ID")
    status = Column(Integer, nullable=False, default=0, comment="商机状态：0:跟进中, 1:已赢单, 2:已输单")
    approval_phase = Column(String(20), nullable=False, default=ApprovalPhase.DRAFT, comment="审批流程状态：draft/pending_review/approved/rejected")
    loss_reason = Column(String(500), nullable=True, comment="输单原因")
    actual_amount = Column(Numeric(12, 2), nullable=True, comment="实际成交金额")
    actual_closing_date = Column(Date, nullable=True, comment="实际成交日期")

    creator_id = Column(String(100), nullable=False, comment="创建人")
    created_time = Column(DateTime, nullable=False, default=func.now(), comment="创建时间")
    last_modified_time = Column(DateTime, nullable=False, default=func.now(), onupdate=func.now(), comment="最后修改时间")
    version = Column(Integer, nullable=False, default=1, comment="版本号（乐观锁）")

    contracts = relationship("Contract", back_populates="opportunity", cascade="all, delete-orphan")
    invoice_applications = relationship("InvoiceApplication", back_populates="opportunity", cascade="all, delete-orphan")

    __table_args__ = (
        Index('idx_customer_id', 'customer_id'),
        Index('idx_procurement_stage_id', 'procurement_stage_id'),
        Index('idx_owner_id', 'owner_id'),
        Index('idx_status', 'status'),
        Index('idx_purchase_type', 'purchase_type'),
        Index('idx_license_type', 'license_type'),
        Index('idx_expected_closing_date', 'expected_closing_date'),
        Index('idx_created_time', 'created_time'),
        Index('idx_team_id', 'team_id'),
        {'comment': '商机表'}
    )

File: CRM-Client/src/api/opportunity.ts
import request from '@/utils/request'
import type { PaginatedResponse } from '@/types/pagination'
import {
  OpportunityApiResponseSchema,
  OpportunityListApiResponseSchema,
  OpportunityListItemApiSchema,
  OpportunityOwnerFilterOptionsResponseSchema,
  SalesFunnelDataSchema,
  StageDurationDataSchema
} from '@/schemas/opportunity'
import { z } from 'zod'

export enum OpportunityStatus {
  FOLLOW_UP = 0,
  WON = 1,
  LOST = 2
}

export enum PurchaseType {
  NEW = 'NEW',
  RENEWAL = 'RENEWAL',
  EXPANSION = 'EXPANSION'
}

export enum LicenseType {
  SUBSCRIPTION = 'SUBSCRIPTION',
  PERPETUAL = 'PERPETUAL'
}

export interface SalesStage {
  id: number
  stage_code: string
  stage_name: string
  win_probability: number
  sort_order: number
  is_active: number
  created_time: string
  updated_time: string
}

export interface SalesStageCreate {
  stage_code: string
  stage_name: string
  win_probability: number
  sort_order: number
}

export interface SalesStageUpdate {
  stage_name?: string
  win_probability?: number
  sort_order?: number
  is_active?: number
}

export interface Opportunity {
  id: number
  opportunity_name: string
  customer_id: number
  customer_name?: string
  procurement_method_id: number | null
  procurement_method_info?: {
    id: number
    code: string
    name: string
    stages?: {
      id: number
      stage_code: string
      stage_name: string
      win_probability: number
      sort_order: number
      is_default_start: boolean
      can_skip: boolean
    }[]
  } | null
  total_amount: number
  user_count: number
  unit_price: number
  license_type: LicenseType
  subscription_years: number | null
  purchase_type: PurchaseType
  decision_maker_count: number | null
  expected_closing_date: string
  procurement_stage_id: number | null
  stage_name?: string
  win_probability: number
  current_stage_snapshot?: {
    id: number
    stage_name: string
    win_probability: number
    template_sort_order: number
    template_code?: string | null
    entered_at: string
    exited_at: string | null
    procurement_method?: {
      id: number
      code: string
      name: string
    } | null
  } | null
  owner_id: string
  owner_info?: OwnerInfo
  creator_id: string
  creator_info?: OwnerInfo
  status: OpportunityStatus
  approval_phase: 'draft' | 'pending_review' | 'approved' | 'rejected'
  actual_amount?: number | null
  actual_closing_date?: string | null
  loss_reason?: string | null
  created_time: string
  updated_time: string
  version: number
  customer_info?: {
    id: number
    account_name: string
  }
}

export interface OwnerInfo {
  id: string
  name: string
  avatar_url: string
}

export interface OwnerFilterOption {
  id: string
  name: string
  is_me: boolean
}

export interface OwnerFilterOptionsResponse {
  data: OwnerFilterOption[]
}

export interface SalesStageDetail {
  id: number
  stage_code: string
  stage_name: string
  win_probability: number
  sort_order: number
  description?: string
  is_active: number
  created_time: string
  last_modified_time: string
}

export interface OpportunityCreate {
  opportunity_name?: string
  customer_id: number
  total_amount: number
  user_count: number
  license_type: LicenseType
  subscription_years?: number | null
  purchase_type: PurchaseType
  expected_closing_date: string
  procurement_method_id?: number | null
  procurement_stage_id?: number | null
  owner_id?: string
  decision_maker_count?: number | null
}

export interface OpportunityUpdate {
  opportunity_name?: string
  total_amount?: number
  user_count?: number
  license_type?: LicenseType
  subscription_years?: number | null
  purchase_type?: PurchaseType
  decision_maker_count?: number | null
  expected_closing_date?: string
  procurement_method_id?: number | null
  procurement_stage_id?: number | null
}

export interface OpportunityMoveStageRequest {
  procurement_stage_id: number
}

export interface OpportunityStageUpdate {
  procurement_stage_id: number
}

export interface OpportunityWinRequest {
  actual_amount: number
  actual_closing_date: string
}

export interface OpportunityLossRequest {
  loss_reason: string
}

export interface OpportunityListParams {
  skip?: number
  limit?: number
  status?: string | number | null
  status_exclude?: string | number | null
  procurement_stage_id?: number | null
  owner_id?: string | null
  owner_id_exclude?: string | null
  customer_id?: number | null
  keyword?: string | null
  customer_keyword?: string | null
  license_type?: string | null
  license_type_exclude?: string | null
  purchase_type?: string | null
  purchase_type_exclude?: string | null
  stage_name?: string | null
  expected_closing_date_start?: string
  expected_closing_date_end?: string
  order_by?: string
  order_dir?: 'asc' | 'desc'
}

export interface SalesFunnelData {
  stage_id: number
  stage_name: string
  opportunity_count: number
  total_amount: number
  average_amount: number
  win_probability: number
}

export interface StageDurationData {
  stage_id: number
  stage_name: string
  avg_duration_days: number
  min_duration_days: number
  max_duration_days: number
  opportunity_count: number
}

export interface OpportunityListResponse {
  id: number
  opportunity_name: string
  customer_id: number
  procurement_method_id: number | null
  procurement_method_info: {
    id: number
    name: string
    description: string | null
    is_active: number
  } | null
  total_amount: number
  user_count: number
  unit_price: number
  license_type: string
  subscription_years: number | null
  purchase_type: string
  decision_maker_count: number | null
  expected_closing_date: string
  stage_id: number | null
  stage_name?: string
  win_probability: number
  owner_id: string
  creator_id: string
  current_stage_snapshot?: {
    id: number
    stage_name: string
    win_probability: number
    template_sort_order: number
    template_code: string
    entered_at: string
    exited_at: string | null
    procurement_method: {
      id: number
      code: string
      name: string
    }
  } | null
  stage: {
    id: number
    stage_code: string
    stage_name: string
    win_probability: number
    sort_order: number
    description: string | null
    is_active: number
    created_time: string
    last_modified_time: string
  } | null
  stage_info: {
    id: number
    stage_name: string
    win_probability: number
    is_default: number
  } | null
  status: number
  approval_phase: 'draft' | 'pending_review' | 'approved' | 'rejected'
  created_time: string
  last_modified_time: string
  customer_name?: string
  owner_info?: {
    id: string
    name: string
    avatar_url?: string | null
  }
}

export const opportunityApi = {
  getOpportunities: async (params?: OpportunityListParams): Promise<OpportunityListResponse[] | PaginatedResponse<OpportunityListResponse>> => {
    // eslint-disable-next-line crmwolf/require-zod-schema
    const response = await request.get<OpportunityListResponse[] | PaginatedResponse<OpportunityListResponse>>(`/v1/opportunities/`, { params })
    return OpportunityListApiResponseSchema.parse(response) as OpportunityListResponse[] | PaginatedResponse<OpportunityListResponse>
  },

  getOpportunity: async (opportunityId: number): Promise<Opportunity> => {
    // eslint-disable-next-line crmwolf/require-zod-schema
    const response = await request.get<Opportunity>(`/v1/opportunities/${opportunityId}`)
    return OpportunityApiResponseSchema.parse(response) as Opportunity
  },

  getOpportunityDetail: async (opportunityId: number): Promise<Opportunity> => {
    // eslint-disable-next-line crmwolf/require-zod-schema
    const response = await request.get<Opportunity>(`/v1/opportunities/${opportunityId}`)
    return OpportunityApiResponseSchema.parse(response) as Opportunity
  },

  createOpportunity: async (data: OpportunityCreate): Promise<Opportunity> => {
    // eslint-disable-next-line crmwolf/require-zod-schema
    const response = await request.post<Opportunity>(`/v1/opportunities/`, data)
    return OpportunityApiResponseSchema.parse(response) as Opportunity
  },

  updateOpportunity: async (opportunityId: number, data: OpportunityUpdate): Promise<Opportunity> => {
    // eslint-disable-next-line crmwolf/require-zod-schema
    const response = await request.put<Opportunity>(`/v1/opportunities/${opportunityId}`, data)
    return OpportunityApiResponseSchema.parse(response) as Opportunity
  },

  moveOpportunityStage: async (opportunityId: number, data: OpportunityMoveStageRequest): Promise<null> => {
    // eslint-disable-next-line crmwolf/require-zod-schema
    const response = await request.post<null>(`/v1/opportunities/${opportunityId}/move-stage`, data)
    return z.null().parse(response)
  },

  markAsWon: async (opportunityId: number, data: OpportunityWinRequest): Promise<null> => {
    const response = await request.patch<null>(`/v1/opportunities/${opportunityId}/win`, data)
    return z.null().parse(response)
  },

  markAsLost: async (opportunityId: number, data: OpportunityLossRequest): Promise<null> => {
    const response = await request.patch<null>(`/v1/opportunities/${opportunityId}/lose`, data)
    return z.null().parse(response)
  },

  deleteOpportunity: async (opportunityId: number): Promise<null> => {
    // eslint-disable-next-line crmwolf/require-zod-schema
    const response = await request.delete<null>(`/v1/opportunities/${opportunityId}`)
    return z.null().parse(response)
  },

  getSalesFunnel: async (): Promise<SalesFunnelData[]> => {
    // eslint-disable-next-line crmwolf/require-zod-schema
    const response = await request.get<SalesFunnelData[]>(`/v1/opportunities/sales-funnel`)
    return z.array(SalesFunnelDataSchema).parse(response) as SalesFunnelData[]
  },

  getStageDuration: async (): Promise<StageDurationData[]> => {
    // eslint-disable-next-line crmwolf/require-zod-schema
    const response = await request.get<StageDurationData[]>(`/v1/opportunities/stage-duration`)
    return z.array(StageDurationDataSchema).parse(response) as StageDurationData[]
  },

  getAvailableForContract: async (customerId: number): Promise<OpportunityListResponse[]> => {
    // eslint-disable-next-line crmwolf/require-zod-schema
    const response = await request.get<OpportunityListResponse[]>(`/v1/opportunities/available-for-contract`, {
      params: { customer_id: customerId }
    })
    return z.array(OpportunityListItemApiSchema).parse(response) as OpportunityListResponse[]
  },

  getOwnerFilterOptions: async (): Promise<OwnerFilterOptionsResponse> => {
    // eslint-disable-next-line crmwolf/require-zod-schema
    const response = await request.get<OwnerFilterOptionsResponse>('/v1/filter-options/owners', { params: { resource: 'opportunity' } })
    return OpportunityOwnerFilterOptionsResponseSchema.parse(response) as OwnerFilterOptionsResponse
  }
}
